ql/src/java/org/apache/hadoop/hive/ql/parse/rewrite/MergeRewriter.java

```diff
@@ -0,0 +1,274 @@
+import org.apache.hadoop.hive.ql.parse.rewrite.sql.SqlGeneratorFactory;
+import org.apache.hadoop.hive.ql.session.SessionState;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+
+public class MergeRewriter implements Rewriter<MergeStatement>, MergeStatement.DestClausePrefixSupplier {
+
+  private final Hive db;
+  protected final HiveConf conf;
+  private final SqlGeneratorFactory sqlGeneratorFactory;
+
+  public MergeRewriter(Hive db, HiveConf conf, SqlGeneratorFactory sqlGeneratorFactory) {
+    this.db = db;
+    this.conf = conf;
+    this.sqlGeneratorFactory = sqlGeneratorFactory;
+  }
+
+  @Override
+  public ParseUtils.ReparseResult rewrite(Context ctx, MergeStatement mergeStatement) throws SemanticException {
+
+    setOperation(ctx);
+    MultiInsertSqlGenerator sqlGenerator = sqlGeneratorFactory.createSqlGenerator();
+    handleSource(mergeStatement.hasWhenNotMatchedInsertClause(), mergeStatement.getSourceAlias(),
+        mergeStatement.getOnClauseAsText(), sqlGenerator);
+
+    MergeStatement.MergeSqlGenerator mergeSqlGenerator = createMergeSqlGenerator(mergeStatement, sqlGenerator);
+    for (MergeStatement.WhenClause whenClause : mergeStatement.getWhenClauses()) {
+      whenClause.toSql(mergeSqlGenerator);
+    }
+
+    boolean validateCardinalityViolation = mergeStatement.shouldValidateCardinalityViolation(conf);
+    if (validateCardinalityViolation) {
+      handleCardinalityViolation(mergeStatement.getTargetAlias(), mergeStatement.getOnClauseAsText(), sqlGenerator);
+    }
+
+    ParseUtils.ReparseResult rr = ParseUtils.parseRewrittenQuery(ctx, sqlGenerator.toString());
+    Context rewrittenCtx = rr.rewrittenCtx;
+    ASTNode rewrittenTree = rr.rewrittenTree;
+    setOperation(rewrittenCtx);
+
+    //set dest name mapping on new context; 1st child is TOK_FROM
+    int insClauseIdx = 1;
+    for (MergeStatement.WhenClause whenClause : mergeStatement.getWhenClauses()) {
+      List<Context.DestClausePrefix> prefixes = whenClause.getDestClausePrefix(this);
+      for (Context.DestClausePrefix prefix : prefixes) {
+        rewrittenCtx.addDestNamePrefix(insClauseIdx, prefix);
+        insClauseIdx++;
+      }
+    }
+
+    if (validateCardinalityViolation) {
+      //here means the last branch of the multi-insert is Cardinality Validation
+      rewrittenCtx.addDestNamePrefix(rewrittenTree.getChildCount() - 1, Context.DestClausePrefix.INSERT);
+    }
+
+    return rr;
+  }
+
+  protected MergeWhenClauseSqlGenerator createMergeSqlGenerator(
+      MergeStatement mergeStatement, MultiInsertSqlGenerator sqlGenerator) {
+    return new MergeWhenClauseSqlGenerator(conf, sqlGenerator, mergeStatement);
+  }
+
+  private void handleSource(boolean hasWhenNotMatchedClause, String sourceAlias, String onClauseAsText,
+                            MultiInsertSqlGenerator sqlGenerator) {
+    sqlGenerator.append("FROM\n");
```

Comment:
that won't work for COW: see https://github.com/apache/hive/pull/4852/files#diff-5612014abb89e13c290474798741828a8933edda57f9035f0656df47c4b63e7bR471